app/Routes/API.js

```diff
@@ -1,91 +1,391 @@
-Router.map(function() {
-	this.route("vmixCharacter", {
-		path: "/vmix-character/:_id/",
-		where: "server",
-		action: function() {
-			this.response.setHeader("Content-Type", "application/json");
-			var query = this.params.query;
-			var key = query && query.key;
-			ifKeyValid(key, this.response, "vmixCharacter", () =>
-				this.response.end(vMixCharacter(this.params._id))
-			);
-		},
-	});
-	this.route("vmixParty", {
-		path: "/vmix-party/:_id/",
-		where: "server",
-		action: function() {
-			this.response.setHeader("Content-Type", "application/json");
-			var query = this.params.query;
-			var key = query && query.key;
-			ifKeyValid(key, this.response, "vmixParty", () =>
-				this.response.end(vMixParty(this.params._id))
-			);
-		},
-	});
-
-	this.route("jsonCharacterSheet", {
-		path: "/character/:_id/json",
-		where: "server",
-		action: function() {
-			this.response.setHeader("Content-Type", "application/json");
-			var query = this.params.query;
-			var key = query && query.key;
-			ifKeyValid(key, this.response, "jsonCharacterSheet", () => {
-				if (canViewCharacter(this.params._id, userIdFromKey(key))){
-					this.response.end(JSONExport(this.params._id))
-				} else {
-					this.response.writeHead(403, "You do not have permission to view this character");
-					this.response.end();
-				}
-			}
-			);
-		},
-	});
+Router.map(function () {
+    this.route("vmixCharacter", {
+        path: "/vmix-character/:_id/",
+        where: "server",
+        action: function () {
+            this.response.setHeader("Content-Type", "application/json");
+            var query = this.params.query;
+            var key = query && query.key;
+            ifKeyValid(key, this.response, "vmixCharacter", () =>
+                this.response.end(vMixCharacter(this.params._id))
+            );
+        },
+    });
+    this.route("vmixParty", {
+        path: "/vmix-party/:_id/",
+        where: "server",
+        action: function () {
```

Comment:
Actions should each just end up being `Meteor.call('someMethod', args)` with all the validation in there, which will be considered trusted code, and will always be allowed

Reply:
Currently to do the validation in a Meteor method I'd have to pass either the Router or Response object to the Meteor method, which throws a RangeError (is it trying to serialize the object to EJSON?). Are there any alternates to this, or do I have to rewrite all the validation code?

Reply:
Traceback: 
```
W20190207-16:39:23.935(-8)? (STDERR) RangeError: Maximum call stack size exceeded
W20190207-16:39:23.937(-8)? (STDERR)     at Array.forEach (<anonymous>)
W20190207-16:39:23.938(-8)? (STDERR)     at Object.EJSON.clone.v [as clone] (packages/ejson/ejson.js:594:18)
W20190207-16:39:23.938(-8)? (STDERR)     at Object.keys.forEach.key (packages/ejson/ejson.js:595:22)
[...repeats a few times]
W20190207-16:39:23.946(-8)? (STDERR)  => awaited here:
W20190207-16:39:23.947(-8)? (STDERR)     at Promise.await (/Users/andrew/.meteor/packages/promise/.0.11.1.1calf8t.hslz++os+web.browser+web.browser.legacy+web.cordova/npm/node_modules/meteor-promise/promise_server.js:60:12)
W20190207-16:39:23.947(-8)? (STDERR)     at Server.apply (packages/ddp-server/livedata_server.js:1634:14)
W20190207-16:39:23.947(-8)? (STDERR)     at Server.call (packages/ddp-server/livedata_server.js:1603:17)
W20190207-16:39:23.947(-8)? (STDERR)     at ctor.<anonymous> (Routes/API.js:73:16)
W20190207-16:39:23.948(-8)? (STDERR)     at boundNext (packages/iron_middleware-stack.js:408:31)
W20190207-16:39:23.948(-8)? (STDERR)     at runWithEnvironment (packages/meteor.js:1356:24)
W20190207-16:39:23.948(-8)? (STDERR)     at packages/meteor.js:1369:14
W20190207-16:39:23.948(-8)? (STDERR)     at ctor.urlencodedParser (/Users/andrew/.meteor/packages/iron_router/.1.1.2.1kmiw8w.53mm++os+web.browser+web.cordova/npm/node_modules/body-parser/lib/types/urlencoded.js:77:44)
W20190207-16:39:23.949(-8)? (STDERR)     at packages/iron_router.js:886:36
W20190207-16:39:23.949(-8)? (STDERR)     at Meteor.EnvironmentVariable.EVp.withValue (packages/meteor.js:1304:12)
W20190207-16:39:23.949(-8)? (STDERR)     at ctor.hookWithOptions (packages/iron_router.js:885:27)
W20190207-16:39:23.950(-8)? (STDERR)     at boundNext (packages/iron_middleware-stack.js:408:31)
W20190207-16:39:23.950(-8)? (STDERR)     at runWithEnvironment (packages/meteor.js:1356:24)
```